### PR TITLE
Fix building NativeShims on OSX

### LINKED ISSUE #76: Can't build  Yubico.NativeShims on osx-arm64
The error is here: [ 28%] Building C object CMakeFiles/Yubico.NativeShims.dir/ssl.bignum.c.o
/Users/nicolasthomas/src/Yubico.NET.SDK/Yubico.NativeShims/ssl.bignum.c:9:1: error: cannot combine with previous 'type-name' declaration specifier
void Native_BN_new(void)
^
1 error generated. 

See fulll logs bellow :
Yubico.NativeShims % sh build-macOS.sh 
~/vcpkg ~/src/Yubico.NET.SDK/Yubico.NativeShims
Already on 'master'
Your branch is up to date with 'origin/master'.
Already up to date.
warning: neither `vcpkg.json` nor `vcpkg-configuration.json` exist to update.
~/src/Yubico.NET.SDK/Yubico.NativeShims
-- Running vcpkg install
Detecting compiler hash for triplet arm64-osx...
Detecting compiler hash for triplet x64-osx...
The following packages will be built and installed:
    openssl[core]:x64-osx -> 3.2.1#2
  * vcpkg-cmake[core]:arm64-osx -> 2023-05-04
  * vcpkg-cmake-config[core]:arm64-osx -> 2022-02-06#1
  * vcpkg-cmake-get-vars[core]:arm64-osx -> 2023-12-31
Additional packages (*) will be modified to complete this operation.
Restored 4 package(s) from /Users/nicolasthomas/.cache/vcpkg/archives in 471 ms. Use --debug to see more details.
Installing 1/4 vcpkg-cmake:arm64-osx...
Elapsed time to handle vcpkg-cmake:arm64-osx: 2.924 ms
Installing 2/4 vcpkg-cmake-get-vars:arm64-osx...
Elapsed time to handle vcpkg-cmake-get-vars:arm64-osx: 1.837 ms
Installing 3/4 vcpkg-cmake-config:arm64-osx...
Elapsed time to handle vcpkg-cmake-config:arm64-osx: 1.653 ms
Installing 4/4 openssl:x64-osx...
Elapsed time to handle openssl:x64-osx: 51.84 ms
Total install time: 529.3 ms
The package openssl is compatible with built-in CMake targets:

    find_package(OpenSSL REQUIRED)
    target_link_libraries(main PRIVATE OpenSSL::SSL OpenSSL::Crypto)

-- Running vcpkg install - done
-- The C compiler identification is AppleClang 15.0.0.15000309
-- The CXX compiler identification is AppleClang 15.0.0.15000309
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test HAVE_STACK_PROTECTOR_ALL
-- Performing Test HAVE_STACK_PROTECTOR_ALL - Success
-- -fstack-protector-all support detected
Checking for PCSC with Mac linkage
PCSC_WINSCARD_H_FOUND: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/PCSC.framework/Headers/winscard.h
HAVE_PCSC_WINSCARD_H: ON
PCSC_MACOSX_LIBS: -Wl,-framework -Wl,PCSC
PCSC_LIBRARIES: -Wl,-framework -Wl,PCSC
PCSC_CFLAGS: 
BACKEND_PCSC: ON
HAVE_PCSC_WINSCARD_H: ON
-- Found OpenSSL: /Users/nicolasthomas/src/Yubico.NET.SDK/Yubico.NativeShims/build64/vcpkg_installed/x64-osx/lib/libcrypto.a (found version "3.2.1")
INFOIPO is enabled.
-- Configuring done (4.6s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/nicolasthomas/src/Yubico.NET.SDK/Yubico.NativeShims/build64
[ 14%] Building C object CMakeFiles/Yubico.NativeShims.dir/pcsc.c.o
[ 28%] Building C object CMakeFiles/Yubico.NativeShims.dir/ssl.bignum.c.o
/Users/nicolasthomas/src/Yubico.NET.SDK/Yubico.NativeShims/ssl.bignum.c:9:1: error: cannot combine with previous 'type-name' declaration specifier
void Native_BN_new(void)
^
1 error generated.
gmake[2]: *** [CMakeFiles/Yubico.NativeShims.dir/build.make:90: CMakeFiles/Yubico.NativeShims.dir/ssl.bignum.c.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/Yubico.NativeShims.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2
cp: ./build64/libYubico.NativeShims.dylib: No such file or directory
-- Running vcpkg install
Detecting compiler hash for triplet arm64-osx...
The following packages will be built and installed:
    openssl[core]:arm64-osx -> 3.2.1#2
  * vcpkg-cmake[core]:arm64-osx -> 2023-05-04
  * vcpkg-cmake-config[core]:arm64-osx -> 2022-02-06#1
  * vcpkg-cmake-get-vars[core]:arm64-osx -> 2023-12-31
Additional packages (*) will be modified to complete this operation.
Restored 4 package(s) from /Users/nicolasthomas/.cache/vcpkg/archives in 431.9 ms. Use --debug to see more details.
Installing 1/4 vcpkg-cmake:arm64-osx...
Elapsed time to handle vcpkg-cmake:arm64-osx: 2.503 ms
Installing 2/4 vcpkg-cmake-get-vars:arm64-osx...
Elapsed time to handle vcpkg-cmake-get-vars:arm64-osx: 3.575 ms
Installing 3/4 vcpkg-cmake-config:arm64-osx...
Elapsed time to handle vcpkg-cmake-config:arm64-osx: 2.219 ms
Installing 4/4 openssl:arm64-osx...
Elapsed time to handle openssl:arm64-osx: 59.69 ms
Total install time: 499.9 ms
The package openssl is compatible with built-in CMake targets:

    find_package(OpenSSL REQUIRED)
    target_link_libraries(main PRIVATE OpenSSL::SSL OpenSSL::Crypto)

-- Running vcpkg install - done
-- The C compiler identification is AppleClang 15.0.0.15000309
-- The CXX compiler identification is AppleClang 15.0.0.15000309
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test HAVE_STACK_PROTECTOR_ALL
-- Performing Test HAVE_STACK_PROTECTOR_ALL - Success
-- -fstack-protector-all support detected
Checking for PCSC with Mac linkage
PCSC_WINSCARD_H_FOUND: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/PCSC.framework/Headers/winscard.h
HAVE_PCSC_WINSCARD_H: ON
PCSC_MACOSX_LIBS: -Wl,-framework -Wl,PCSC
PCSC_LIBRARIES: -Wl,-framework -Wl,PCSC
PCSC_CFLAGS: 
BACKEND_PCSC: ON
HAVE_PCSC_WINSCARD_H: ON
-- Found OpenSSL: /Users/nicolasthomas/src/Yubico.NET.SDK/Yubico.NativeShims/buildarm/vcpkg_installed/arm64-osx/lib/libcrypto.a (found version "3.2.1")
INFOIPO is enabled.
-- Configuring done (3.0s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/nicolasthomas/src/Yubico.NET.SDK/Yubico.NativeShims/buildarm
[ 14%] Building C object CMakeFiles/Yubico.NativeShims.dir/pcsc.c.o
[ 28%] Building C object CMakeFiles/Yubico.NativeShims.dir/ssl.bignum.c.o
/Users/nicolasthomas/src/Yubico.NET.SDK/Yubico.NativeShims/ssl.bignum.c:9:1: error: cannot combine with previous 'type-name' declaration specifier
void Native_BN_new(void)
^
1 error generated.
gmake[2]: *** [CMakeFiles/Yubico.NativeShims.dir/build.make:90: CMakeFiles/Yubico.NativeShims.dir/ssl.bignum.c.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/Yubico.NativeShims.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2
cp: ./buildarm/libYubico.NativeShims.dylib: No such file or directory

### PR DESCRIPTION
**Pull request by AdamVe**
# Description

Updates the source files so that it is possible to build macos shims for arm64 and x86_64 platforms.

Fixes #76  

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

The build which fails in main/develop does not fail anymore.

- [x] My code follows the [style guidelines](https://raw.githubusercontent.com/Yubico/Yubico.NET.SDK/043119ad1d19e0e6e66556c970a81d0c1aba36c8/CONTRIBUTING.md) of this project 
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings

### CODE REVIEW

File: Yubico.NativeShims/build-macOS.sh [hunk 1]
```diff
@@ -7,13 +7,13 @@ git pull
...
-\cmake \
+cmake \
     -S . \
     -B ./build64 \
     -DCMAKE_BUILD_TYPE=Release \
     -DCMAKE_TOOLCHAIN_FILE=$VCPKG_INSTALLATION_ROOT/scripts/buildsystems/vcpkg.cmake \
-    -DVCPKG_TARGET_TRIPLET=x64-osx
-
+    -DVCPKG_TARGET_TRIPLET=x64-osx \
+    -DCMAKE_OSX_ARCHITECTURES=x86_64
```
Comment (GregDomzalski) [inline Yubico.NativeShims/build-macOS.sh:16]: At one point we used custom toolchain files. It had set the `CMAKE_OSX_ARCHITECTURE` option among others. We probably unintentionally stopped using them when we started using VCPKG.

I'd recommend taking a look at the ones we have in the `cmake/` subfolder. We should probably either start using them as overlays (I think that's the term vcpkg uses) or get rid of them.

Either way - would probably be good to check to see if there are other settings that were used once upon a time that might be worth reintroducing.

I'd also double check to see if `VCPKG_TARGET_TRIPLET` doesn't already set/imply `CMAKE_OSX_ARCHITECTURES` if you haven't done so already.

Comment (AdamVe) [reply Yubico.NativeShims/build-macOS.sh:16]: Good points! I did some investigations here and this is what I found out:

Although the triplets set `VCPKG_OSX_ARCHITECTURES` to correct values, these is not used at all on OSX leaving `CMAKE_OSX_ARCHITECTURES` undefined, which will fail linkage for specific build/host/target combinations. Explicitly defining the value for each target platform fixes the issue.

I was not aware of the `cmake/osx*cmake` files, currently they are not used and in the original commit they served as the main toolchain files. They could be reused through `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` but I don’t see any advantage to just using `-DCMAKE_OSX_ARCHITECTURES` as vcpkg is used now.

See also that the `cmake/osx*cmake` files do cache the specific `CMAKE_OSX_ARCHITECTURES` value (functionally same effect as using the explicit `-D`).

My way forward would be to use `-DCMAKE_OSX_ARCHITECTURES=arch` and delete the unused cmake files. What do you think?

Comment (GregDomzalski) [reply Yubico.NativeShims/build-macOS.sh:16]: Yes, that sounds quite reasonable to me. Thanks Adam!

**Review by DennisDyallo — APPROVED** [review]
LGTM. Adding void as the parameter clarifies the intent of the method and since we're more strict on compiler warnings I can see how this would help.

File: Yubico.NativeShims/build-macOS.sh (hunk 1 above)
Comment (DennisDyallo) [reply Yubico.NativeShims/build-macOS.sh:16]: +1 for simplifying and deleting unused files. Thanks @AdamVe !

**Pull request merged**